lib/streamlit/elements/widgets/radio.py

```diff
@@ -51,46 +53,89 @@
 @dataclass
 class RadioSerde(Generic[T]):
     options: Sequence[T]
-    index: int
+    index: int | None
+
+    def serialize(self, v: object) -> int | None:
+        if v is None:
+            return None
 
-    def serialize(self, v: object) -> int:
-        if len(self.options) == 0:
-            return 0
-        return index_(self.options, v)
+        return 0 if len(self.options) == 0 else index_(self.options, v)
 
     def deserialize(
         self,
-        ui_value: Optional[int],
+        ui_value: int | None,
         widget_id: str = "",
-    ) -> Optional[T]:
+    ) -> T | None:
         idx = ui_value if ui_value is not None else self.index
 
         return (
             self.options[idx]
-            if len(self.options) > 0 and self.options[idx] is not None
+            if idx is not None
+            and len(self.options) > 0
+            and self.options[idx] is not None
             else None
         )
 
 
 class RadioMixin:
-    @gather_metrics("radio")
+    @overload
     def radio(
         self,
         label: str,
         options: OptionSequence[T],
         index: int = 0,
         format_func: Callable[[Any], Any] = str,
-        key: Optional[Key] = None,
-        help: Optional[str] = None,
-        on_change: Optional[WidgetCallback] = None,
-        args: Optional[WidgetArgs] = None,
-        kwargs: Optional[WidgetKwargs] = None,
+        key: Key | None = None,
+        help: str | None = None,
+        on_change: WidgetCallback | None = None,
+        args: WidgetArgs | None = None,
+        kwargs: WidgetKwargs | None = None,
+        *,  # keyword-only args:
+        disabled: bool = False,
+        horizontal: bool = False,
+        captions: Sequence[str] | None = None,
+        label_visibility: LabelVisibility = "visible",
+    ) -> T:
```

Comment:
Hm, is the return type here correct? I think there's a case where `st.radio` is passed an empty options sequence and unspecified index (which thus defaults to 0) where the return type of the call to `st.radio` will be `None`.

This will probably be fine most of the time as `T` will often be unbound, but a user could have explicitly typed the `OptionsSequence` doing something like the following

```python
my_options: List[str] = []

# Some conditional logic which doesn't append anything to `my_options`
# In this script run
if False:
    my_options.append("option 1")

choice = st.radio('pick an option', my_options)
```

In the code snippet above, I think `mypy` will complain that a `str` is not returned by
`st.radio`, but I would consider the above code valid.

I think the other overload + the signature used in the function implementation itself should be enough to cover all cases.

Reply:
Oh, you are right. Empty option list also returns `None` and there doesn't seem to be a good way to type an empty list. So, I think we can just remove the overload since it always needs to be typed with `None` & the type of the options.